### PR TITLE
Initial commit. Add set and get methods for applicable attributes.

### PR DESCRIPTION
**Pull request by katcavers**
## Description

This PR is the first step in bringing generated API methods to the model level for issue #962 . This commit includes changes to add a get and set method for applicable attributes at the model level. Currently, these additional methods are skipped over in the code generation stage, so generation of API methods remains unchanged.

I tried to find all the generators and add the necessary code to skip generation of the get and set methods. If any were missed, please let me know.

## Tests

Added a test in UmpleParserTest called autoGeneratedAPIMethods.

### CODE REVIEW

**Comment by vahdat-ab** [comment]
@katcavers The build is failing. Please fix them and request your PR again.

**Comment by katcavers** [comment]
@vahdat-ab It is fixed now. I had forgotten to commit one file, oops!

File: cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java [hunk 1]
```diff
@@ -2560,6 +2560,70 @@ public void methodNameConflictUsingDefaulted()
     assertFailedParse("045_phpDefaultedNameConflict.ump", 48);
     assertFailedParse("045_rubyDefaultedNameConflict.ump", 48);
   }
+  
+  @Test
+  public void autoGeneratedAPIMethods()
```
Comment (vahdat-ab) [inline cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java]: It is better to keep test cases simple.  Each test case should be dealing with a specific case. It helps later to explore the test case easier.  One test case for every stereotype. You can keep the source model the same for all test cases.

**Review by vahdat-ab — CHANGES_REQUESTED** [review]

**Comment by vahdat-ab** [comment]
Great. 
You also need to take care of errors like 1009. The way they have been implemented must be changed because now we have auto-generated methods at the model.
you can get more details about errors related to attributes in `en.error`.
you can create another issue for the metioned case and I'll merge this PR, or you can update the PR and then I'll merge. I prefer the second one :-)

File: cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java (hunk 1 above)
Comment (katcavers) [reply cruise.umple/test/cruise/umple/compiler/UmpleParserTest.java]: Is there a way to only have to call assertSimpleParse once if I separate them all into separate tests? Or do I have to call it now for each test?

**Comment by vahdat-ab** [comment]
@katcavers the build is failing

**Comment by katcavers** [comment]
@vahdat-ab yes, I know, I didn't mean to push the last commit. I commented on your code review, but perhaps you didn't see it? Just in case, here is the question again:
Is there a way to only have to call assertSimpleParse once if I separate them all into separate tests? Or do I have to call it now for each test?
Thanks!

**Comment by vahdat-ab** [comment]
Yes, you can do it inside `public void setUp()`. However, I would suggest breaking your file to several files in the direct connection with the tests you do. In that case, you parse things (use assertSimpleParse ) related to each test case.

**Comment by katcavers** [comment]
Okay, so should I (1) keep the tests where they are in UmpleParserTest and create separate .ump files to parse for each test case, or (2) create a new test file for these tests and use the setup method to parse the .ump file once, or (3) create a new test file for these tests and use separate .ump files for each of them?

Which would be best? :)

**Comment by vahdat-ab** [comment]
Option 1

File: cruise.umple/src/Generator_CodeGvClassDiagram.ump [hunk 2]
```diff
@@ -129,6 +129,9 @@ class GvClassDiagramGenerator
     if (hasSuboption("showmethods")) {
       boolean isFirstMethod = true;
       for (Method uMethod : inUmpleClassifier.getMethods()) {
+      	if (uMethod.getSource() == Method.Source.fAutoAPI)
+      	  continue;
+
```
Comment (vahdat-ab) [inline cruise.umple/src/Generator_CodeGvClassDiagram.ump:134]: I wonder why you wanted to escape these methods here?
Actually, it's a nice way to show what them in the model. By default, we don't shows any methods, but when they are selected they should be shown

**Comment by vahdat-ab** [comment]
@TimLethbridge I wonder if you wanted to have Umple auto-generated methods to appear in generators like Json, scxml, etc. To me, it makes sense to have them there as well.

File: cruise.umple/src/Generator_CodeGvClassDiagram.ump (hunk 2 above)
Comment (vahdat-ab) [reply cruise.umple/src/Generator_CodeGvClassDiagram.ump:134]: as we discussed online, disregard my point.

**Comment by vahdat-ab** [comment]
I wonder if you have done sth related to this PR or not? If no, so I want to merge this.
Probably you can add remove methods to this PR so they are all merged together.

**Comment by katcavers** [comment]
Hey Vahdat, yes I had thought that after our meeting a week ago it was going to be pulled. But you are right, adding the remove methods to this PR makes sense. Hopefully I have that ready to go by the end of today.

**Comment by katcavers** [comment]
Hi Vahdat, I saw one of the checks failed, and looked in the logs to find this:
    [junit] Running cruise.statemachine.test.DoActivityTest
    [junit] Tests run: 8, Failures: 0, Errors: 1, Time elapsed: 36.028 sec

I believe this is why it failed? I'm not sure how I can debug the problem? All tests passed locally
@vahdat-ab

**Comment by katcavers** [comment]
I'm also really confused as to why this commit could cause this to start failing...the only difference is a new test and two methods which are actually never called anywhere in the code yet. The previous commit passed both travis-ci builds.

**Comment by vahdat-ab** [comment]
That's ok. The test fails because of pool/queue state machines. This is a random failure.

File: cruise.umple/src/Umple_Code.ump [hunk 3]
```diff
@@ -1966,6 +1966,41 @@ class UmpleClass
...
+  	String firstLetter = attributeName.substring(0,1).toUpperCase();
+  	String rest = "";
+  	if (attributeName.length() > 1)
+  	{
+  		rest = attributeName.substring(1);
+  	}
+  	String getName = "get" + firstLetter + rest;
+  	String setName = "set" + firstLetter + rest;
+  	
+  	ArrayList<Method> methodsCopy = new ArrayList<Method>(this.getMethods());
```
Comment (vahdat-ab) [inline cruise.umple/src/Umple_Code.ump]: It is better to define the variable more general.
`List<Method> methodsCopy = new ArrayList<Method>(this.getMethods());`

**Review by vahdat-ab — COMMENTED** [review]
I have a question:
do yo deal with constructor at the modeling level or not?
if yes then when an attribute is removed its constructor needs to be modified.

**Comment by katcavers** [comment]
Yes, I am working on the constructor now. I am almost done, but I had not considered this. I will look into it tomorrow. Thank you!

File: cruise.umple/src/Umple_Code.ump (hunk 3 above)
Comment (katcavers) [reply cruise.umple/src/Umple_Code.ump]: Okay I will make this change.

**Pull request merged**